lib/jsonapi-serializers/serializer.rb

```diff
@@ -320,19 +320,20 @@ def self.serialize_primary(object, options = {})
       return if object.nil?
 
       serializer = serializer_class.new(object, options)
-      data = {
+
+      {
         'id' => serializer.id.to_s,
         'type' => serializer.type.to_s,
         'attributes' => serializer.attributes,
-      }
-
-      # Merge in optional top-level members if they are non-nil.
-      # http://jsonapi.org/format/#document-structure-resource-objects
-      data.merge!({'attributes' => serializer.attributes}) if !serializer.attributes.nil?
-      data.merge!({'links' => serializer.links}) if !serializer.links.nil?
-      data.merge!({'relationships' => serializer.relationships}) unless serializer.relationships.empty?
-      data.merge!({'meta' => serializer.meta}) if !serializer.meta.nil?
-      data
+      }.tap do |data|
```

Comment:
I guess I assumed that `merge!` basically does a `h = h.merge(key => value)`.  I ran a `benchmark-ips` test to see what the performance looks like:

```ruby
require 'benchmark/ips'

Benchmark.ips do |x|
  x.config(:time => 5, :warmup => 2)

  x.report("merge") do
    h = {1 => 2}
    h.merge(3 => 4)
  end

  x.report("merge!") do
    h = {1 => 2}
    h.merge!(3 => 4)
  end

  x.report("key assignment") do
    h = {1 => 2}
    h[1] = 2
  end

  x.compare!
end
```

The results:

```
Calculating -------------------------------------
               merge    19.472k i/100ms
              merge!    35.164k i/100ms
      key assignment    54.054k i/100ms
-------------------------------------------------
               merge    295.619k (±15.3%) i/s -      1.441M
              merge!    624.239k (± 8.7%) i/s -      3.130M
      key assignment      1.132M (± 8.2%) i/s -      5.676M

Comparison:
      key assignment:  1132317.5 i/s
              merge!:   624239.2 i/s - 1.81x slower
               merge:   295619.1 i/s - 3.83x slower
```

So it seems like key assignment is the fastest, followed by `merge!` which is 1.81x slower, followed by `merge` which is 3.83x slower than key assignment.

It's an interesting result, perhaps, but:

 * This says nothing about memory usage
 * This test is testing tens of thousands of iterations per second.  In this case we're talking about four assignments of keys to the hash so it's probably not a huge deal for CPU or memory

Even if the `merge!` is kept, I might suggest getting rid of the brackets like:

```ruby
data.merge!('links' => serializer.links)
```

I can respect the desire for clear returns ;)

So really, perhaps, this comes down to two things which I see as significant changes:

 * The code is calling the various methods on `serializer` twice each time which does unnecessary calculations
 * There's no need to assign `serializer.attributes` to the hash twice

This could be written as six lines with three pairs of:

 * assignment to a variable
 * assignment to the hash key (or a `merge!` if the value isn't blank)

So it's questionable if the loop provides value, especially since it's just three iterations instead of four if we leave out `attributes`.  Having it all splayed out could be clearer, but one other reason that I tend to like making things like this loops is that it's a declaration that you're doing the same thing over and over.  If you copy/paste/edit the code then a person has to look carefully at each line to see that they're doing the same thing